test/local/senders/sms.js

```diff
@@ -4,77 +4,183 @@
 
-  it('sends a valid sms without a signinCode', () => {
-    return sms.send('+442078553000', 'installFirefox', 'en')
-      .then(() => {
-        assert.equal(publish.callCount, 1, 'AWS.SNS.publish was called once')
-        assert.equal(publish.args[0].length, 1, 'AWS.SNS.publish was passed one argument')
-        assert.deepEqual(publish.args[0][0], {
+      it('called sns.getSMSAttributes correctly', () => {
+        assert.equal(sns.getSMSAttributes.callCount, 1)
+        const args = sns.getSMSAttributes.args[0]
+        assert.equal(args.length, 1)
+        assert.deepEqual(args[0], { attributes: [ 'MonthlySpendLimit' ] })
+      })
+
+      it('called cloudwatch.getMetricsStatistics correctly', () => {
+        assert.equal(cloudwatch.getMetricsStatistics.callCount, 1)
+        const args = cloudwatch.getMetricsStatistics.args[0]
+        assert.equal(args.length, 1)
+        const now = new Date()
+        assert.equal(args[0].Namespace, 'AWS/SNS')
+        assert.equal(args[0].MetricName, 'SMSMonthToDateSpentUSD')
+        assert.equal(args[0].StartTime, new Date(`${now.getFullYear()}-${now.getMonth() + 1}-01Z`).toISOString())
+      })
+
+      it('isBudgetOk returns true', () => {
+        assert.strictEqual(sms.isBudgetOk(), true)
+      })
+
+      it('did not call sns.publish', () => {
+        assert.equal(sns.publish.callCount, 0)
+      })
+
+      it('did not call log.error', () => {
+        assert.equal(log.error.callCount, 0)
+      })
+    })
+
+    describe('spend > threshold:', () => {
+      beforeEach(() => {
+        results.getMetricsStatistics.Datapoints[1].Sum = 1
+      })
+
+      it('isBudgetOk returns true', () => {
+        assert.strictEqual(sms.isBudgetOk(), true)
+      })
+
+      describe('wait a tick:', () => {
+        beforeEach(done => setImmediate(done))
+
+        it('isBudgetOk returns false', () => {
+          assert.strictEqual(sms.isBudgetOk(), false)
+        })
+
+        it('did not call log.error', () => {
+          assert.equal(log.error.callCount, 0)
+        })
+      })
+    })
+
+    describe('invalid data:', () => {
+      beforeEach(() => {
+        results.getMetricsStatistics.Datapoints[1].Sum = 'wibble'
+      })
+
+      describe('wait a tick:', () => {
+        beforeEach(done => setImmediate(done))
+
+        it('isBudgetOk returns true', () => {
+          assert.strictEqual(sms.isBudgetOk(), true)
+        })
+
+        it('called log.error correctly', () => {
+          assert.equal(log.error.callCount, 1)
+          const args = log.error.args[0]
+          assert.equal(args.length, 1)
+          assert.deepEqual(args[0], {
+            op: 'sms.budget.error',
+            err: 'Invalid Datapoints'
+          })
+        })
+      })
+    })
+
+    describe('send a valid sms without a signinCode:', () => {
+      beforeEach(() => {
+        return sms.send('+442078553000', 'installFirefox', 'en')
+      })
+
+      it('called sns.publish correctly', () => {
```

Comment:
The other thing that was annoying me was how these tests were doing their set-up inside the test and just lumping all of the assertions together in a big ugly lump. Thus de-lumped.